InterfacePython/testOptimization.py

```diff
@@ -0,0 +1,112 @@
+    def __init__(self):
+        unittest.TestCase.__init__(self, "testNothing")
+
+    def testNothing(self):
+        pass
+
+    def checkParticleSwarmState(self):
+        iNumDimensions = 2
+        iNumParticles = 5
+        # Initialization tests.
+        state = Opt.ParticleSwarmState(iNumDimensions, iNumParticles)
+        self.assertEqual(iNumDimensions, state.getNumDimensions())
+        self.assertEqual(iNumParticles, state.getNumParticles())
+        self.assertFalse(state.isPositionInitialized())
+        self.assertFalse(state.isVelocityInitialized())
+        self.assertFalse(state.isBestPositionInitialized())
+        self.assertFalse(state.isCacheInitialized())
+        # Loading/Unloading tests.
+        pp = np.random.rand(iNumParticles, iNumDimensions)
+        state.setParticlePositions(pp)
+        self.assertTrue(state.isPositionInitialized())
+        self.assertTrue(np.array_equal(pp, state.getParticlePositions()))
+        pv = np.random.rand(iNumParticles, iNumDimensions)
+        state.setParticleVelocities(pv)
+        self.assertTrue(np.array_equal(pv, state.getParticleVelocities()))
+        bpp = np.random.rand(iNumParticles+1, iNumDimensions)
+        state.setBestParticlePositions(bpp)
+        self.assertTrue(state.isBestPositionInitialized())
+        self.assertTrue(np.array_equal(bpp, state.getBestParticlePositions()))
+        state.clearBestParticles()
+        self.assertFalse(state.isBestPositionInitialized())
+        self.assertTrue(np.array_equal(np.zeros([iNumParticles + 1, iNumDimensions]), state.getBestParticlePositions()))
+        # Generation tests.
+        state = Opt.ParticleSwarmState(iNumDimensions, iNumParticles)
+        state.initializeParticlesInsideBox(np.array([-1.0, 1.0]), np.array([2.0, 3.0]))
+        self.assertTrue(state.isPositionInitialized())
+        self.assertTrue(state.isVelocityInitialized())
+        pp = state.getParticlePositions()
+        self.assertTrue(np.all((-1.0 <= pp[:,0]) & (pp[:,0] <= 2.0)))
+        self.assertTrue(np.all((1.0 <= pp[:,1]) & (pp[:,1] <= 3.0)))
+
+    def checkParticleSwarm(self):
+        iNumDimensions = 2
+        iNumParticles = 50
+        state = Opt.ParticleSwarmState(iNumDimensions, iNumParticles)
+        state.initializeParticlesInsideBox(np.array([-3.0, -2.0]), np.array([3.0, 2.0]))
+        # Six-hump-camel function.
+        shc = lambda x : (4 - 2.1*x[0]*x[0] + x[0]*x[0]*x[0]*x[0]/3)*x[0]*x[0] + x[0]*x[1] + (-4.0 + 4.0*x[1]*x[1])*x[1]*x[1]
+        f = lambda x_batch : np.apply_along_axis(shc, 1, x_batch)
+        inside = lambda x : bool((-3 <= x[0]) and (x[0] <= 3) and (-2 <= x[1]) and (x[1] <= 2))
+        # Main call + tests.
+        Opt.ParticleSwarm(f, 1, inside, state, 0.5, 2, 2)
+        self.assertTrue(state.isCacheInitialized())
+        state.clearCache()
+        self.assertFalse(state.isCacheInitialized())
+        iNumIterations = 200
+        Opt.ParticleSwarm(f, iNumIterations, inside, state, 0.5, 2, 2)
```

Comment:
How do you control for randomness here. We must fix the random seed, otherwise this will fail periodically.